From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Agner <stefan@agner.ch>
Date: Mon, 8 Aug 2022 13:39:26 +0200
Subject: [PATCH] Incremental update device permission

---
 update.go | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/update.go b/update.go
index 4f22064a..117c1ae8 100644
--- a/update.go
+++ b/update.go
@@ -362,10 +362,7 @@ other options are ignored.
 		config.Cgroups.Resources.Unified = r.Unified
 
 		if len(r.Devices) > 0 {
-			config.Cgroups.Resources.Devices = nil
-			defaultAllowedDevices := specconv.CreateDefaultDevicesCgroups(&config)
-
-			err = specconv.CreateCgroupDeviceConfig(config.Cgroups.Resources, &r, defaultAllowedDevices)
+			err = specconv.CreateCgroupDeviceConfig(config.Cgroups.Resources, &r, nil)
 			if err != nil {
 				return err
 			}
